File: src/design_patterns/creational/factory_method/c#/factory_method.cs
using System;

// Product
abstract class Animal
{
    public abstract string Speak();
}

// Concrete Products
class Dog : Animal
{
    public override string Speak()
    {
        return "Woof!";
    }
}

class Cat : Animal
{
    public override string Speak()
    {
        return "Meow";
    }
}

// Creator
abstract class AnimalFactory
{
    public abstract Animal CreateAnimal();
}

// Concrete Creators
class DogFactory : AnimalFactory
{
    public override Animal CreateAnimal()
    {
        return new Dog();
    }
}

class CatFactory : AnimalFactory
{
    public override Animal CreateAnimal()
    {
        return new Cat();
    }
}

// Client code
class Client
{
    static void Main()
    {
        AnimalFactory factory;
        Animal animal;

        factory = new DogFactory();
        animal = factory.CreateAnimal();
        Console.WriteLine(animal.Speak());  // Output: "Woof!"

        factory = new CatFactory();
        animal = factory.CreateAnimal();
        Console.WriteLine(animal.Speak());  // Output: "Meow"
    }
}

File: src/design_patterns/creational/factory_method/python/factory_method.py
from abc import ABC, abstractmethod


class Animal(ABC):
    """ """
    @abstractmethod
    def speak(self):
        """ """
        pass


class Dog(Animal):
    """ """
    def speak(self):
        """ """
        return "Woof!"


class Cat(Animal):
    """ """
    def speak(self):
        """ """
        return "Meow"


class AnimalFactory:
    """ """
    def create_animal(self, animal_type):
        """

        :param animal_type: 

        """
        if animal_type == "Dog":
            return Dog()
        elif animal_type == "Cat":
            return Cat()
        else:
            return None


factory = AnimalFactory()
dog = factory.create_animal("Dog")
print(dog.speak())  # Output: "Woof!"
cat = factory.create_animal("Cat")
print(cat.speak())  # Output: "Meow"
